### DISCUSSION TITLE
Enabling Course Consumption Flow through a third party app

### DISCUSSION THREAD

**Discussion by kameshbhr** [Sunbird-Ed/Community#28] [Documentation] [answered]
Hi,

We are planning to open up DIKSHA APIs for ecosystem players to enable building solutions on top of DIKSHA.  One of the use cases is to enable third party apps allow users to view and consume Courses that are published on DIKSHA. Towards this, I am trying to create documentation around the set of APIs that are involved in the end to end flow of the Course Consumption. I've looked at the microsites of various Sunbird BBs and based on my understanding came up with the following list:
https://docs.google.com/spreadsheets/d/1IpZV49G_HHZw8PRXyLpNsoR90EvrTpnXsKXovvLTWhc

There are some missing pieces in it which I couldn't figure out. I have marked them in red colour.  Can someone please validate this and help me filling the missing pieces?

Thanks,
Kamesh

**Comment by surendrasinghs** [comment]
Today the Course Assessments are played using the Content Player component in Sunbird Knowlg. When a third-party/external app plays the Course Assessment will it be able to generate a minimal set of telemetry events (and raise/post it through Telemetry service) for the subsequent Course progress and other data products to use them successfully? 
The most critical event is the summary event which is posted using Content State update API `course/v1/content/state/update`

The app and portal enforce maximum attempts for a Course Assessment. Will the third-party honour the 'maxAttempts' attribute of a Content and behave accordingly? Maximum attempts are fetched using Content state read `course/v1/content/state/read`

Will the third-party/external app use a Sunbird player to play the Course Assessment or will it try to build/leverage some other player? Using a Sunbird player might give a headstart rather than trying to use/build another player. Integrating a Sunbird player might need some effort if it's not already compatible with certain technology (e.g. React, Native Android, iOS, etc).

@sharathkashyap @vinukumar-vs @pallakartheekreddy please chime-in.

**Reply by Krishnaj20** [reply]
@kameshbhr - This question is not specific to your query but to understand more about the possible use case landscape - will the ecosystem use their collection player to render the experience or will they rely on Sunbird players?

**Reply by kameshbhr** [reply]
@surendrasinghs @Krishnaj20 Thanks for your response.
To your questions on whether third-party apps use the Sunbird players or not, it would be their choice. We would only provide the various options.

From your response, here is what I understand. Please let me know if it is correct:
1. Sunbird player has the logic of calculating the Course assessment score. I assume the Sunbird player also has logic to send the score to the server through telemetry.
2.  If a third-party app wants to play Course assessments, the two options are: a) Build their own player where the player has to take care of score calculation and sending it to the server through telemetry.  b) Integrate Sunbird player in the app

Follow up questions:
1. Does the Sunbird player also have the logic of restricting a user to max number of attempts (as defined in the Course assessment)?
2. What is the telemetry even to send the Course assessment score back to server?

**Reply by pramodkvarma** [reply]
Course metadata should drive player behavior. Any hard coding or assumptions on the side of player is a bad idea. Of course, player can have the capability, but, the actual enabling/disabling/flow-control/etc should be controlled via player config (that can be programmatically injected when instantiating player) plus content metadata.

**Reply by vinukumar-vs** [reply]
@kameshbhr @pramodkvarma 
Content-player is not sending any data directly to the server. It sends data to the parent(mobile/portal) who is launching the player. 

In the case of course assessment, the parent application(mobile/portal) is sending configuration to the sunbird-player to show a custom end-page(because it is a course assessment) which will show the "submit" button to the user. Once the user confirms to submit the score, the parent application is sending the telemetry data(assess events) to the server.

So any third-party apps that want to implement the same, have to pass the configuration to the player for course assessment to show the custom endpage(part of player itself, has default & custom endpage specific to assessment). Third party app should have the logic to capture all assessEvents generated by the player. On player submit event, third party app should send this data to their server.

@surendrasinghs @rhwarrier

**Reply by pramodkvarma** [reply]
OK, makes sense. Thanks for clarifying @vinukumar-vs

**Reply by kameshbhr** [reply]
@vinukumar-vs Thanks for the details. This clarifies how app can use player config and how the assessment data is sent to the server.

Looking forward for clarifications related to the other API in the Course consumption flow.

**Comment by kameshbhr** [comment]
@luckynvdu @rhwarrier @sharathkashyap Can you please validate the APIs in the Course Consumption flow and provide inputs on the missing pieces in the spreadsheet that I shared.

Also, few more questions popped up regarding how does API handle scenarios like:
- Same user trying to enroll to the same batch twice
- User tries to enroll to a Batch that is closed or enrollment date is crossed
- Can someone game the course completion in their app using the API - say the Course progress is updated to 100% without course materials being completed?

**Reply by pramodkvarma** [reply]
I am curious on the question on gaming! Are you saying one individual end user (learner) can write an app for themselves, connect to production, and manipulate their own scores? In most platforms, that's not even possible given app builders are typically institutions and are not learners themselves.

**Reply by kameshbhr** [reply]
Not a user, but an organization can potentially do that right? I am trying to understand the checks and balances that need to be put in when opening these APIs for a larger ecosystem. What can be taken care by the APIs and what needs to be taken care through process.

**Reply by pramodkvarma** [reply]
What's their incentive to offer cheating to their own users?

**Reply by kameshbhr** [reply]
If an org gets monetary or other kind of rewards (say from their funder or a funding authority) based on the number of certifications/course completions - that could become an incentive. I am trying to play a devil's advocate here.

**Reply by pramodkvarma** [reply]
OK, understood. Given APIs, entire learning can be fudged. Important not to fund the way you are saying, but, to fund based on some sort of assessment (on another channel).

**Reply by kameshbhr** [reply]
Ok, got it. Thanks.

**Comment by sharathkashyap** [comment, marked as answer]
@kameshbhr I have tried to provide my answers in the comments section for some of the questions posted in the sheet.
cc: @luckynvdu

**Reply by kameshbhr** [reply]
Thanks @sharathkashyap 
Few questions that are still open:
- What is the API for user login?
- What is the API to store the user consent?
- For calculating and updating overall progress of a Course, you mentioned - "Progress is computed on both front end and backend separately.". Is there an API that needs to be explicitly called from front end to update the Course progress to the backend?
- What is the telemetry event to submit the assessment score?

**Reply by sharathkashyap** [reply]
@kameshbhr Please find the responses below

`What is the API for user login?`
This is a Key Cloak based Login which calls Portal Backend API to create a key-cloak session. Will try to add more information in the sheet attached.

`What is the API to store the user consent?`
learner/user/v1/consent/update is the API we use as part of consent update.

`For calculating and updating overall progress of a Course, you mentioned - "Progress is computed on both front end and backend separately.". 

Is there an API that needs to be explicitly called from front end to update the Course progress to the backend?`
Course Progress is calculated by /content/course/v1/content/state/read. 

What is the telemetry event to submit the assessment score?
We will send group of ASSESS Events as part of content state update. With each question represented by 1 ASSESS Event of telemetry.

**Comment by sharathkashyap** [comment]
@kameshbhr Locking this thread due to inactivity.
